examples/deepspeed-multinode/sky.yaml

```diff
@@ -0,0 +1,80 @@
+# Example: a distributed DeepSpeed job (DeepSpeed-Chat) on 2 VMs.
+#
+# This takes care constructing a "hostfile" to pass to DeepSpeed.
+#
+# Usage:
+#
+#   $ sky launch sky.yaml -r --down -c ds
+#
+#   # After the job starts running, log into the two nodes and check gpustat:
+#
+#   $ ssh ds
+#   $ gpustat -i
+#
+#   $ ssh ds-worker1
+#   $ gpustat -i
+
+resources:
+  accelerators: T4:1  # AWS, Azure, GCP
+  # accelerators: A100:1  # GCP, Lambda
+  # accelerators: A100-80GB:1  # Azure, GCP, SCP
+num_nodes: 2
+
+setup: |
+  set -ex
+  git clone https://github.com/microsoft/DeepSpeedExamples.git || true
+  cd DeepSpeedExamples
+  git checkout d7c42b4f34df91035e7ed3e0c51500bb53d0bc71
+
+  conda create -n deepspeed python=3.8 -y
+  source activate deepspeed
+
+  pip install deepspeed
+
+  cd applications/DeepSpeed-Chat
+  pip install -r requirements.txt
+
+  # Required by DeepSpeed in multi-node settings.
+  sudo apt-get -y install pdsh
+
+file_mounts:
+  # Required for DeepSpeed's passwordless SSH (run commands on nodes).
+  ~/.ssh/id_rsa: ~/.ssh/sky-key
+
+run: |
+  cd DeepSpeedExamples
+  source activate deepspeed
+
+  # Launch on the first node only
+  if [ "${SKYPILOT_NODE_RANK}" == "0" ]; then
+
+    # Prepare a hostfile.
+    HOSTFILE_PATH=/tmp/hostfile.${SKYPILOT_TASK_ID}
+    python3 -c "import os;n_gpus=os.environ['SKYPILOT_NUM_GPUS_PER_NODE'];print('\n'.join([f'{ip} slots={n_gpus}' for ip in os.environ['SKYPILOT_NODE_IPS'].splitlines()]))" > ${HOSTFILE_PATH}
```

Comment:
nit: for readability, maybe we can do it in multiple lines:
```
python -c <<-EOF
  import os
  num_gpus = os.environ['SKYPILOT_NUM_GPUS_PER_NODE']
  configs = [f'{ip} slots={num_gpus} for IP in os.environ['SKYPILOT_NODE_IPS'].splitlines()]
  with open(os.environ['HOSTFILE_PATH'], 'w') as f:
    f.write('\n'.join(configs))
EOF
```

Reply:
It seems the following works for me:
```
run: |
  export HOSTFILE_PATH=/tmp/hostfile.${SKYPILOT_TASK_ID}
  python <<EOF

  import os
  num_gpus = os.environ['SKYPILOT_NUM_GPUS_PER_NODE']
  configs = [f'{ip} slots={num_gpus}' for ip in os.environ['SKYPILOT_NODE_IPS'].splitlines()]
  with open(os.environ['HOSTFILE_PATH'], 'w') as f:
    f.write('\n'.join(configs))
  
  EOF

  echo bye
  cat $HOSTFILE_PATH

```

Reply:
For some reason putting this into the full YAML doesn't work, but a more minimal YAML works: https://gist.github.com/concretevitamin/f98c77b1688f2c635d6d7e0a37b1cb44

Reply:
Thanks for the testing! I think the problem is caused by the indent of the if block. Due to the limitation of python, we have to do something like the following:
```
if [ <condition> ]; then
    echo hi
    python <<EOF
import os
if True:
    print("Hello, World!")
    print("Your argument is: " + os.environ['ARG'])
EOF
    echo bye
fi
```

I don't feel strongly for its readability. I will let you decide whether to change. ; )

Reply:
Will keep one-liner for now. We can handle it for user if there's demand.